File: content/snippets/scala2/interview_lcci/03/05_sort_stack.scala
object Lcci0305SortStack {
  def sortStack(st: scala.collection.mutable.ArrayBuffer[Int]): Unit = {
    val tmp = scala.collection.mutable.ArrayBuffer.empty[Int]
    while (st.nonEmpty) {
      val cur = st.remove(st.length - 1)
      while (tmp.nonEmpty && tmp.last > cur) {
        st += tmp.remove(tmp.length - 1)
      }
      tmp += cur
    }
    while (tmp.nonEmpty) {
      st += tmp.remove(tmp.length - 1)
    }
  }

  def main(args: Array[String]): Unit = {
    val st = scala.collection.mutable.ArrayBuffer(3, 1, 4, 2)
    sortStack(st)
    println(st.reverse.mkString(" "))
  }
}

File: content/snippets/python/interview_lcci/03/05_sort_stack.py
def sort_stack(st: list[int]) -> None:
    tmp: list[int] = []
    while st:
        cur = st.pop()
        while tmp and tmp[-1] > cur:
            st.append(tmp.pop())
        tmp.append(cur)
    while tmp:
        st.append(tmp.pop())


def main() -> None:
    st = [3, 1, 4, 2]
    sort_stack(st)
    print(" ".join(str(x) for x in reversed(st)))


if __name__ == "__main__":
    main()
